src/Datadog.Trace/AsyncLocalScopeManager.cs

```diff
@@ -8,26 +10,50 @@ internal class AsyncLocalScopeManager : IScopeManager
 
         private readonly AsyncLocalCompat<Scope> _activeScope = new AsyncLocalCompat<Scope>();
 
+        public event EventHandler<ScopeEventArgs> ScopeOpened;
+
+        public event EventHandler<ScopeEventArgs> ScopeActivated;
+
+        public event EventHandler<ScopeEventArgs> ScopeDeactivated;
+
+        public event EventHandler<ScopeEventArgs> ScopeClosed;
+
         public Scope Active => _activeScope.Get();
 
         public Scope Activate(Span span, bool finishOnClose)
         {
             var activeScope = _activeScope.Get();
             var scope = new Scope(activeScope, span, this, finishOnClose);
+
+            var scopeDeactivatedArgs = new ScopeEventArgs(activeScope);
+            var scopeOpenedArgs = new ScopeEventArgs(scope);
+
+            ScopeOpened?.Invoke(this, scopeOpenedArgs);
+
             _activeScope.Set(scope);
+            ScopeDeactivated?.Invoke(this, scopeDeactivatedArgs);
```

Comment:
Should we deactivate if there is no active scope prior to this call?

Reply:
Good catch, we should not be sending an event in this case. Let me fix that.